Update EIP-2780: clarify 2780 (#11891)
ethereum/EIPs: EIPS/eip-2780.md

Once the transaction passes the intrinsic gas check, the state-dependent charges are applied at runtime — as the first frame is entered — drawing from the same gas the transaction supplies. Like every charge in this EIP, each runtime charge is split into regular gas and state gas per [EIP-8037](./eip-8037.md): account accesses and writes are regular gas, while durable state growth (new account leaves and net-new delegation bytes) is state gas. This mirrors how a `CALL` frame is metered: accesses, account creation, and writes are charged as state is touched, not up front. Because these charges require state access, they are evaluated only after [EIP-7702](./eip-7702.md) authorizations are applied.

> Before any runtime charge is applied, `accessed_addresses` is initialized at transaction start under the pre-existing rules, unchanged: the sender, `tx.to`(if not `None`), the precompiles ([EIP-2929](./eip-2929.md)), the access-list entries ([EIP-2930](./eip-2930.md)), and the coinbase ([EIP-3651](./eip-3651.md)).

Reviewer: Imo, this does not need clarification as we dont change EIP-2929/EIP-3651 behaviour.

@misilva73 if you think this is worth adding, i would move it later in EIP not in Abstract

Author: I am fine to remove this clarification, but i would prefer to remove the `and added to accessed_addresses` in runtime gas charging. It confuses me a lot as the to is already pre-warmed with pre-amsterdam rules.

Reviewer 2: I tend to agree. I would not put it in the abstract. Happy to move it to the spec sections if you think it helps with clarity

While processing each valid authorization — the per-authority rules of [EIP-8037](./eip-8037.md) ensure each charge below is applied at most once per authority:

- if the authority account does not already exist, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas for the new account leaf;

> - if this is the first write to the authority within the transaction, charge `ACCOUNT_WRITE` in regular gas; the charge is skipped when the authority's write is already paid for: as `tx.sender` (by `TX_BASE_COST`), or as `tx.to` of a value-bearing transaction (by `TX_VALUE_COST`);

Reviewer: So this is a custom rule, this feels unnecessary, and it clashes with other parts. 

`tx.sender` nonce and balance should already be changed before applying the auth item, so first `if this is the first write to the authority` stands.

`tx.to` check for `ACCOUNT_WRITE` is charged after auth list, so when applying delegation, it is charged, but when doing tx.to check account would exist, so no `ACCOUNT_WRITE` is going to charged.

Author: > tx.sender nonce and balance should already be changed before applying the auth item, so first if this is the first write to the authority stands.

Yes, true. Maybe I can change the words in this way:

`if this is the first write to the authority within the transaction, charge `ACCOUNT_WRITE` in regular gas`. The first write actually covers the scenarios I mentioned.

Author: > tx.to check for ACCOUNT_WRITE is charged after auth list, so when applying delegation, it is charged, but when doing tx.to check account would exist, so no ACCOUNT_WRITE is going to charged.

Sorry I don't understand this reply.

What I meant is: if authority == tx.to, and if we haven't charged `ACCOUNT_WRITE` for tx.to (e.g, value == 0, in this case, we only charge Cold access in IntrinsicGas), then we need to charge the `ACCOUNT_WRITE` in auth processing.

But true, as you said, `the first write to the authority within the transaction` also covers this scenario. 

I am wondering if these clarifications in detail will be useful for implementors. If you guys think it's not, i can also remove it

Reviewer: It is not clarifications that you are proposing here, it is custom rules. 

if we skip `tx.to` charge in eip7702 but change the account, later in path we need to check if `tx.to` exists and charge `ACCOUNT_WRITE`, in this example `tx.to` would exist but we would have skipped eip7702 charge. So it is not correct

Author: Maybe my words is confusing, what I am trying to specify is:

In these cases, the authority has already been charged and the `Account-write` shouldn't be applied one more time:
- authority == tx.sender
- authority has been written by preceding valid authorizations
- authority == tx.to and tx.value != zero

It's all the cases I can imagine that `Account-write` has been applied for this authority and so i express these words as "clarifications"

Author: Anyway, what do you think I change the sentence as below?

`- if this is the first write to the authority within the transaction, charge ACCOUNT_WRITE in regular gas;`

`i.e. the authority differs from tx.to` this part is not correct and confusing. Differs from tx.to doesn't mean  the authority is modified for the first time.

Reviewer 2: I think we should make this clearer. I think it would help to explicitly add int he EIP the three cases @rjl493456442 listed, i.e.:

- authority == tx.sender
- authority has been written by preceding valid authorizations
- authority == tx.to and tx.value != zero

- if the authorization writes the 23-byte delegation indicator into a previously empty slot (net-new delegation bytes), charge `STATE_BYTES_PER_AUTH_BASE × CPSB` in state gas.

`REGULAR_PER_AUTH_BASE_COST` already prices the authority access at the cold rate at the intrinsic phase, so no account-access charge is added at runtime for the authority; the authority is added to `accessed_addresses`, per [EIP-7702](./eip-7702.md).

> Then the recipient account is loaded — its access was already charged at the cold rate at the intrinsic phase, and it is warm from transaction start — and unless the transaction is a self-transfer or a contract creation:

Reviewer: Do agree that `and added to `accessed_addresses`` needs removal. For impl, we could do warming here as warm/cold part was never checked before.

Bigger question is should this be a place where the recipient is added to BAL?

I would do something like
```suggestion
Then the recipient account is added to the BAL - its access was already charged at the cold rate at the intrinsic phase, and it is warm from the transaction start.
```

wdyt?

Author: > Bigger question is should this be a place where the recipient is added to BAL?

Yes it should.

Here we check the callee existence as a part of runtime gas charging. The account should definitely be added into BAL due to database lookup.

Author: done

While processing each valid authorization — the per-authority rules of [EIP-8037](./eip-8037.md) ensure each charge below is applied at most once per authority:

- if the authority account does not already exist, charge `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` in state gas for the new account leaf;
- if this is the first write to the authority within the transaction, charge `ACCOUNT_WRITE` in regular gas; the charge is skipped when the write is already paid for, namely when the authority:
  - is `tx.sender`, covered by `TX_BASE_COST`;
  - was written by a preceding valid authorization;

>   - is `tx.to` of a value-bearing transaction, covered by `TX_VALUE_COST`;

Reviewer 3: `TX_VALUE_COST` is 4244 and the `ACCOUNT_WRITE` is 8000. So I'm not sure `tx.to` is covered. Might be worth looking into.

Author: https://hackmd.io/@bFEBbZiVSAO0IURh9qzEFg/

Yes, but I think it's a technical debt, maybe @misilva73 has some insights. At least in the EIP description, it says TX_VALUE_COST is for recipient balance update and value transfer

Reviewer 2: TX_VALUE_COST was derived in two steps:
- First, we take the diff between the cost of a transfer without value (15k) and a transfer with value (21k). This is based on [benchmarks](https://misilva73.github.io/eip-2780-repricing/index.html). Note that Erigon is still working on their optimizations, so I am taking the 2nd worst client for now.
- To this diff, we remove the log cost - this cost covers mostly history growth, so the execution cost is orthogonal.

- if the authorization writes the 23-byte delegation indicator into a previously empty slot (net-new delegation bytes), charge `STATE_BYTES_PER_AUTH_BASE × CPSB` in state gas.

`REGULAR_PER_AUTH_BASE_COST` already prices the authority access at the cold rate at the intrinsic phase, so no account-access charge is added at runtime for the authority; the authority is added to `accessed_addresses`, per [EIP-7702](./eip-7702.md).

> Then the recipient account is added to the BAL — its access was already charged at the cold rate at the intrinsic phase, and it is warm from transaction start — and unless the transaction is a self-transfer or a contract creation:

Reviewer 3: This slightly contradicts EIP-7928 as it currently stands. It says
"Transaction sender and recipient addresses (even for zero-value transfers)" MUST be included in the BAL.

If we include the recipient at this stage, the BAL for a transaction which OOGs at the auth processing will not have the recipient (in contradiction to EIP-7928).

Reviewer: This was true for EIP-7928 where it can't halt in runtime. This EIP changes this behaviour and it makes sense that is changes BAL.

> Then the recipient account is added to the BAL — its access was already charged at the cold rate at the intrinsic phase, and it is warm from transaction start — and unless the transaction is a self-transfer or a contract creation:

Reviewer 3: Also, the `not self-transfer` gate should not be applied here. Self-transfer to an account which already has a 7702 delegation should be charged for accessing the delegation. This is already in the table as "ETH transfer to self, sender 7702 delegated". We must make this text consistent with the table

Author: Sorry Guru, I don't get your point here. 

For self-transfer, we need to charge the delegation-target access if it has delegation, and it's explicitly specified below.

I don't get why it's relevant with `not self-transfer` gate. Can you elaborate it?

Reviewer 3: It may just be an issue with how I am reading it. The way it is currently structured, it seems to say

```
Then the recipient account is added to the BAL and unless the transaction is a self-transfer or a contract creation:

- if the recipient is non-existent and tx.value > 0, charge STATE_BYTES_PER_NEW_ACCOUNT × CPSB in state gas;
- if the recipient is an [EIP-7702](https://github.com/rjl493456442/EIPs/blob/2780-clarify/EIPS/eip-7702.md) delegated account, additionally charge the delegation-target access in regular gas — COLD_ACCOUNT_ACCESS if cold, WARM_ACCESS if warm.
```

Which makes me think the 2 charges should not apply if it is a self transfer. In any case, this is very minor

- **[EIP-2930](./eip-2930.md) (access lists).** Access lists keep their existing per-entry charges and warming semantics for execution-level touches. Listing the recipient or an authority does not reduce the intrinsic cold-rate charges; it only pre-warms the address for execution.
- **[EIP-7623](./eip-7623.md) (calldata floor).** EIP-7623 floors a transaction's cost at `21,000 + TOTAL_COST_FLOOR_PER_TOKEN × tokens_in_calldata`, where the flat `21,000` is the intrinsic base this EIP decomposes. With this EIP active, that base term is replaced by the transaction's decomposed regular-gas intrinsic — the sum of the regular-gas intrinsic primitives (`TX_BASE_COST + COLD_ACCOUNT_ACCESS + TX_VALUE_COST + TRANSFER_LOG_COST` for value transfers, `TX_BASE_COST + CREATE_ACCESS` for contract creation) — so the floor and its validity check rest on the same per-transaction base as the rest of intrinsic gas rather than a stale constant. Like the floor, this base is state-independent, so the state-dependent runtime charges (the new-account state-gas charge and the delegation-target access) do not enter it. The calldata schedule itself is unchanged; only the base the floor sits on moves.

> - **[EIP-7702](./eip-7702.md) (set EOA code).** `TX_BASE_COST` is unchanged even when the sender temporarily assumes code; clients must not perform a disk code load to classify the sender, since EIP-7702 provides the code inline. When `tx.to` is a delegated account, resolving the delegation loads the target's code, charged `COLD_ACCOUNT_ACCESS` on first touch or warm thereafter, following the standard EIP-2929 `accessed_addresses` model. Setting a delegation (writing the 23-byte pointer) is distinct from resolving one (reading the target's code); the resolution charge is genuine work and is not prepaid by the authorization. Authorization cost itself is decomposed: `REGULAR_PER_AUTH_BASE_COST` — which already includes the authority's cold access — is charged per authorization at the intrinsic phase, while the state-dependent charges (`STATE_BYTES_PER_NEW_ACCOUNT × CPSB` for a non-existent authority, `ACCOUNT_WRITE` for the first write to the authority within the transaction, and `STATE_BYTES_PER_AUTH_BASE × CPSB` for net-new delegation bytes) are charged at runtime during authorization processing, replacing the worst-case `PER_EMPTY_ACCOUNT_COST` intrinsic charge and refund.

Reviewer: github is not great at showing this diff. The change is:

`STATE_BYTES_PER_NEW_ACCOUNT × CPSB` plus `ACCOUNT_WRITE` for a non-existent authority, and `STATE_BYTES_PER_AUTH_BASE × CPSB` for net-new delegation bytes

to

`STATE_BYTES_PER_NEW_ACCOUNT × CPSB` for a non-existent authority, `ACCOUNT_WRITE` for the first write to the authority within the transaction, and `STATE_BYTES_PER_AUTH_BASE × CPSB` for net-new delegation bytes

Author: Exactly, the real diff is to decouple the account-creation (STATE_BYTES_PER_NEW_ACCOUNT × CPSB ) with ACCOUNT_WRITE, one stands for state growth, one stands for database write, making no sense to pair them

5. Value transfer to a 7702-delegated account: `21,000` intrinsic + `COLD_ACCOUNT_ACCESS` runtime = `24,000` regular plus execution.
6. Contract-creation transaction, `value = 0`: `23,000` intrinsic regular; `183,600` runtime state if the deployment target did not exist, `0` otherwise.
7. Contract-creation transaction, `value > 0`: `24,756` intrinsic regular; `183,600` runtime state if the deployment target did not exist, `0` otherwise.

> 8. EIP-7702 transaction: `TX_BASE_COST` is unchanged even when the sender assumes code. Each authorization is charged `REGULAR_PER_AUTH_BASE_COST` (which already covers the cold authority access) at the intrinsic phase; at runtime, a non-existent authority additionally pays `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` state, the first write to the authority within the transaction pays `ACCOUNT_WRITE` regular (skipped for `tx.sender`, or for `tx.to` of a value-bearing transaction), and net-new delegation bytes add `STATE_BYTES_PER_AUTH_BASE × CPSB` state — replacing the flat `PER_EMPTY_ACCOUNT_COST` intrinsic charge and refund.

Reviewer: ```suggestion
8. EIP-7702 transaction: `TX_BASE_COST` is unchanged even when the sender assumes code. Each authorization is charged `REGULAR_PER_AUTH_BASE_COST` (which already covers the cold authority access) at the intrinsic phase; at runtime, a non-existent authority additionally pays `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` state, the first write to the authority within the transaction pays `ACCOUNT_WRITE` regular, and net-new delegation bytes add `STATE_BYTES_PER_AUTH_BASE × CPSB` state — replacing the flat `PER_EMPTY_ACCOUNT_COST` intrinsic charge and refund.
```
